community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulator.java

```diff
@@ -0,0 +1,418 @@
+            {
+                fail( descriptor, t );
+            }
+            finally
+            {
+                activeTasks.decrementAndGet();
+            }
+        } );
+    }
+
+    /**
+     * Shuts down the executor waiting {@link #AWAIT_TIMEOUT_MINUTES} minutes for it's termination.
+     * Restores the interrupted status and exits normally when interrupted during waiting.
+     *
+     * @param now <code>true</code> if {@link ExecutorService#shutdownNow()} should be used and <code>false</code> if
+     * {@link ExecutorService#shutdown()} should be used.
+     * @throws IllegalStateException if tasks did not complete in {@link #AWAIT_TIMEOUT_MINUTES} minutes.
+     */
+    private void shutdownExecutor( boolean now )
+    {
+        log.info( (now ? "Forcefully shutting" : "Shutting") + " down executor." + EOL + this );
+        if ( now )
+        {
+            executor.shutdownNow();
+        }
+        else
+        {
+            executor.shutdown();
+        }
+
+        try
+        {
+            boolean tasksCompleted = executor.awaitTermination( AWAIT_TIMEOUT_MINUTES, TimeUnit.MINUTES );
+            if ( !tasksCompleted )
+            {
+                handleTimeout();
+            }
+        }
+        catch ( InterruptedException e )
+        {
+            handleInterrupt();
+        }
+    }
+
+    private void handleTimeout()
+    {
+        throw new IllegalStateException( "Index population tasks were not able to complete in " +
+                                         AWAIT_TIMEOUT_MINUTES + " minutes." + EOL + this + EOL + findStackTraces() );
+    }
+
+    private void handleInterrupt()
+    {
+        Thread.currentThread().interrupt();
+        log.warn( "Interrupted while waiting for index population tasks to complete." + EOL + this );
+    }
+
+    private List<NodePropertyUpdate> newBatch()
+    {
+        return new ArrayList<>( BATCH_SIZE );
+    }
+
+    private ExecutorService createThreadPool()
+    {
+        int threads = Math.max( 2, Runtime.getRuntime().availableProcessors() - 1 );
+        BlockingQueue<Runnable> workQueue = new LinkedBlockingQueue<>( TASK_QUEUE_SIZE );
+        ThreadFactory threadFactory = daemon( FLUSH_THREAD_NAME_PREFIX );
+        RejectedExecutionHandler rejectedExecutionHandler = new ThreadPoolExecutor.CallerRunsPolicy();
+        return new ThreadPoolExecutor( threads, threads, 0L, TimeUnit.MILLISECONDS, workQueue, threadFactory,
+                rejectedExecutionHandler );
+    }
+
+    /**
+     * Finds all threads and corresponding stack traces relevant for this populator which can potentially cause the
+     * {@link ExecutorService executor} to not terminate in {@link #AWAIT_TIMEOUT_MINUTES} minutes. Such threads are
+     * ones used by executor itself and Lucene Merge threads which can cause stalls of writers.
+     *
+     * @return thread dump as string for all relevant threads.
+     */
+    private static String findStackTraces()
```

Comment:
do we need to filter threads here? what if we will get all stack traces instead if we already in that miserable state?